# Guided Project: Answering Business Questions using SQL
- Write a SQL query to extract the relevant data
- Write a short paragraph, drawing conclusions and explaining the data and/or visualizations.



Import the SQLite and pandas.

In [7]:
import sqlite3
import pandas as pd

Create a run_query() function, that takes a SQL query as an argument and returns a pandas dataframe of that query.

In [10]:
def run_query(q):
    with sqlite3.connect('chinook.db') as conn:
        return pd.read_sql(q, conn)

Create a run_command() function that takes a SQL command as an argument and executes it using the sqlite module.

In [12]:
def run_command(c):
    with sqlite3.connect('chinook.db') as conn:
        conn.isolation_level = None
        conn.execute(c)

Create a show_tables() function that calls the run_query() function to return a list of all tables and views in the database.

In [32]:
def show_tables():
    q = '''
        SELECT name, type
        FROM sqlite_master
        WHERE type IN ("table","view");
        '''
    return run_query(q)

In [33]:
show_tables() 

name   type
0            album  table
1           artist  table
2         customer  table
3         employee  table
4            genre  table
5          invoice  table
6     invoice_line  table
7       media_type  table
8         playlist  table
9   playlist_track  table
10           track  table

The record label specializes in artists from the USA, and they have given Chinook some money to advertise the new albums in the USA, so we're interested in finding out which genres sell the best in the USA.

- write a query to find out which genres sell the most tracks in the USA
- create a visualization of that data using pandas.

In [50]:
run_query('''
WITH usa_tracks_sold AS
   (
    SELECT il.* FROM invoice_line il
    INNER JOIN invoice i on il.invoice_id = i.invoice_id
    INNER JOIN customer c on i.customer_id = c.customer_id
    WHERE c.country = "USA"
   )

SELECT
    g.name genre,
    count(uts.invoice_line_id) tracks_sold,
    cast(count(uts.invoice_line_id) AS FLOAT) / (
        SELECT COUNT(*) from usa_tracks_sold
    ) percentage_sold
FROM usa_tracks_sold uts
INNER JOIN track t on t.track_id = uts.track_id
INNER JOIN genre g on g.genre_id = t.genre_id
GROUP BY 1
ORDER BY 2 DESC
LIMIT 10;
        ''')

genre  tracks_sold  percentage_sold
0                Rock          561         0.533777
1  Alternative & Punk          130         0.123692
2               Metal          124         0.117983
3            R&B/Soul           53         0.050428
4               Blues           36         0.034253
5         Alternative           35         0.033302
6               Latin           22         0.020932
7                 Pop           22         0.020932
8         Hip Hop/Rap           20         0.019029
9                Jazz           14         0.013321

Based on the sales of tracks across different genres in the USA, we should purchase the new albums by the following artists:

Red Tone (Punk)
Slim Jim Bites (Blues)
Meteor and the Girls (Pop)
It's worth keeping in mind that combined, these three genres only make up only 17% of total sales, so we should be on the lookout for artists and albums from the 'rock' genre, which accounts for 53% of sales.